File: Whatsapp/OpenWA/sdk/javascript/src/resources/catalog.ts
/**
 * Catalog resource — WhatsApp Business catalog, products, and product/catalog sends.
 *
 * Backed by `src/modules/catalog/catalog.controller.ts` (`@Controller('sessions/:sessionId')`).
 * NOTE: the catalog controller is mounted under the session root, so catalog
 * reads are `/catalog...` while product/catalog SENDS share the messages
 * namespace (`/messages/send-product`, `/messages/send-catalog`). All require
 * an OPERATOR-level key for write operations.
 * @packageDocumentation
 */

import { encodeSegment } from '../http.js';
import type { OpenWAClient } from '../client.js';
import type {
  CatalogInfo,
  CatalogProduct,
  CatalogProductsQuery,
  MessageResponse,
  PaginatedProducts,
  SendCatalogRequest,
  SendProductRequest,
} from '../types.js';

export class CatalogResource {
  constructor(private readonly client: OpenWAClient) {}

  /** Get the business catalog info. */
  info(sessionId: string): Promise<CatalogInfo> {
    return this.client.request<CatalogInfo>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/catalog`,
    });
  }

  /** List catalog products. Returns a `{ products, pagination }` page. */
  products(sessionId: string, query?: CatalogProductsQuery): Promise<PaginatedProducts> {
    return this.client.request<PaginatedProducts>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/catalog/products`,
      query,
    });
  }

  /** Get a single product by id. */
  product(sessionId: string, productId: string): Promise<CatalogProduct> {
    return this.client.request<CatalogProduct>({
      method: 'GET',
      path: `/api/sessions/${encodeSegment(sessionId)}/catalog/products/${encodeSegment(productId)}`,
    });
  }

  /** Send a product message. Requires an OPERATOR-level key. Shares the messages path. */
  sendProduct(sessionId: string, body: SendProductRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-product`,
      body,
    });
  }

  /** Send a catalog link message. Requires an OPERATOR-level key. Shares the messages path. */
  sendCatalog(sessionId: string, body: SendCatalogRequest): Promise<MessageResponse> {
    return this.client.request<MessageResponse>({
      method: 'POST',
      path: `/api/sessions/${encodeSegment(sessionId)}/messages/send-catalog`,
      body,
    });
  }
}

File: OpenWA/sdk/python/openwa/resources/catalog.py
"""Catalog resource — WhatsApp Business catalog, products, and product/catalog sends.

Backed by ``src/modules/catalog/catalog.controller.ts``
(``@Controller('sessions/:sessionId')``). NOTE: the catalog controller is
mounted under the session root, so catalog reads are ``/catalog...`` while
product/catalog SENDS share the messages namespace
(``/messages/send-product``, ``/messages/send-catalog``).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._http import quote_segment
from ..types import (
    CatalogInfo,
    CatalogProduct,
    CatalogProductsQuery,
    MessageResponse,
    PaginatedProducts,
    SendCatalogRequest,
    SendProductRequest,
)

if TYPE_CHECKING:
    from .._http import HttpExecutor


class CatalogResource:
    def __init__(self, http: "HttpExecutor") -> None:
        self._http = http

    def info(self, session_id: str) -> CatalogInfo:
        return self._http.request("GET", f"/api/sessions/{quote_segment(session_id)}/catalog")

    def products(
        self, session_id: str, query: CatalogProductsQuery | None = None
    ) -> PaginatedProducts:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/catalog/products", query=query
        )

    def product(self, session_id: str, product_id: str) -> CatalogProduct:
        return self._http.request(
            "GET", f"/api/sessions/{quote_segment(session_id)}/catalog/products/{quote_segment(product_id)}"
        )

    def send_product(self, session_id: str, body: SendProductRequest) -> MessageResponse:
        """Send a product message. Requires an OPERATOR-level key. Shares the messages path."""
        return self._http.request(
            "POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-product", body=body
        )

    def send_catalog(self, session_id: str, body: SendCatalogRequest) -> MessageResponse:
        """Send a catalog link message. Requires an OPERATOR-level key. Shares the messages path."""
        return self._http.request(
            "POST", f"/api/sessions/{quote_segment(session_id)}/messages/send-catalog", body=body
        )
